README.md
# HSP-booking-automation
An Cronjob automation that automaticly books your sport courses for the Hochschulsport Hamburg. This only works in combination with the HSP "Ab zum Sport Card".

1. It runs 5 min. after the course has started and books for next week.
2. Requires Firefox´s or Chromes geckodriver in bash etc
3. Logs and Errors can be found in `executions.txt`. Crontab output is found in `crontab_log.txt`. 

## Installation
1. `python3 -m venv venv`
2. `source venv/bin/activate`
3. `pip3 install -r requirements.txt`
4. get geckodriver for your OS from here: https://github.com/mozilla/geckodriver/releases. To run the script on Raspberry Pi you may have to use Chromes Selenium Driver (`sudo apt-get install chromium-chromedriver` or `sudo apt-get install chromium-browser`).

### User specific setup
1. Make file called `settings.ini` and write:\
    <code>[Your Name]\
    Sport=[url_to_sport_website_with_booking_button,XPATH_to_booking_button]</code>\
     <em><strong>!!! Change "Your Name" and "Sport"-Value accordingly</strong></em>
2. Make file called `.env` with your credentials:\
<code>EMAIL={"Name":"name@email.com"}\
PASSWORD={"Name":"your_secret_passwd"}</code>\
 <em><strong>!!! Change your Name accordingly</strong></em>
\
*Note: These Passwords get saved as envoirenemt Variables and will not be seen by anyone. As long as you keep your `.env` file secret.*

## Run the automation
Make a cronjob with the command:\
 `35 19 * * 1 sh ~/HSP_Automation/run_booking.sh "Sport" "Your Name"`\
 <em><strong>!!! Change "Your Name" and "Sport"-Value and Crontab-Timing accordingly</strong></em>



Wettersonden_alert/history.txt
#########################################################################
Report from:14:42 11.03.2021

-----Launch facility: Schleswig-----
Start Time: 12UHR
ETA: 13:57:15
Landing Site: 54.684916N 12.239673O 
last heigth: -4.627624
=============

-----Launch facility: Sasel-----
Start Time: 12UHR
ETA: 13:38:56
Landing Site: 54.058743N 12.595946O 
last heigth: 21.535677
=============

-----Launch facility: Norderney-----
Start Time: 12UHR
ETA: 13:38:56
Landing Site: 54.058743N 12.595946O 
last heigth: 21.535677
=============

-----Launch facility: Norderney-----
Start Time: 12UHR
ETA: 13:38:56
Landing Site: 54.058743N 12.595946O 
last heigth: 21.535677
=============

-----Launch facility: Norderney-----
Start Time: 12UHR
ETA: 13:38:56
Landing Site: 54.058743N 12.595946O 
last heigth: 21.535677
=============

-----Launch facility: Bergen-----
Start Time: 12UHR
ETA: 13:38:56
Landing Site: 54.058743N 12.595946O 
last heigth: 21.535677
=============

-----Launch facility: Bergen-----
Start Time: 12UHR
ETA: 13:38:56
Landing Site: 54.058743N 12.595946O 
last heigth: 21.535677
=============

-----Launch facility: Bergen-----
Start Time: 12UHR
ETA: 13:38:56
Landing Site: 54.058743N 12.595946O 
last heigth: 21.535677
=============


=============


Wettersonden_alert/predicted.py
from datetime import datetime
import requests
import json
import math

now = datetime.now()
curdate = now.strftime("%H:%M %d.%m.%Y")
day = now.weekday()
today = now.strftime("%Y-%m-%dT")
intoday = input("enter the date from ehich you want teh predictions (format: Year-Month-Day e.g.(2021-03-04))")
today = intoday+"T"
today.replace(" ", "")
homepos = input("enter Koordinates of the region the flights schould land in: (default is Hamburg (53.551228,9.890442))")
homekoord = [53.551228,9.890442]
homekoord = homepos.split(",")
if len(homekoord) !=2:
    print("your input is invalid! Default is used!")
    homekoord = [53.551228,9.890442]
print("your query is for:",today,"at",homekoord)

data = [
["Schleswig",[5.51,10.27,31308],54.527231,9.548441],
["Sasel",[5.11,10.43,24064],53.649111,10.110911],
["Norderney",[5.33,9.83,32289],53.712111,7.151811],
["Bergen",[5.01,13.06,33511],52.815411,9.924511]]

predicitons = []
landings = []
notlandins = []

for facilities in data:
    url6 = "http://predict.cusf.co.uk/api/v1/?launch_latitude="+str(facilities[2])+"&launch_longitude="+str(facilities[3])+"&launch_altitude=20&launch_datetime="+str(today)+"05:00:00Z&ascent_rate="+str(facilities[1][0])+"&burst_altitude="+str(facilities[1][2])+"&descent_rate="+str(facilities[1][1])
    url12 = "http://predict.cusf.co.uk/api/v1/?launch_latitude="+str(facilities[2])+"&launch_longitude="+str(facilities[3])+"&launch_altitude=20&launch_datetime="+str(today)+"11:00:00Z&ascent_rate="+str(facilities[1][0])+"&burst_altitude="+str(facilities[1][2])+"&descent_rate="+str(facilities[1][1])
    url18 = "http://predict.cusf.co.uk/api/v1/?launch_latitude="+str(facilities[2])+"&launch_longitude="+str(facilities[3])+"&launch_altitude=20&launch_datetime="+str(today)+"17:00:00Z&ascent_rate="+str(facilities[1][0])+"&burst_altitude="+str(facilities[1][2])+"&descent_rate="+str(facilities[1][1])
    urls = [url6,url12,url18]

    if facilities[0] == "Sasel":
        if day == 2 or day == 3:    #wenn Mi oder Do ist dann dartet in Sasel eine Sonde
            try:
                response = requests.get(url12)
            except:
                print("error for Sasel")
            data = json.loads(response.text)
            lastPoint = []
            for x, points in enumerate(data["prediction"][1]["trajectory"]):
                lastPoint = [points["datetime"],points["latitude"],points["longitude"],points["altitude"]]
            predicitons.append([facilities[0],data["request"]["launch_datetime"],lastPoint[0],lastPoint[1],lastPoint[2],lastPoint[3]])

    if facilities[0] == "Schleswig":
        try:
            response = requests.get(url12)
        except:
            print("error for Schleswig")
        data = json.loads(response.text)
        lastPoint = []
        for x, points in enumerate(data["prediction"][1]["trajectory"]):
            lastPoint = [points["datetime"],points["latitude"],points["longitude"],points["altitude"]]
        predicitons.append([facilities[0],data["request"]["launch_datetime"],lastPoint[0],lastPoint[1],lastPoint[2],lastPoint[3]])


    if facilities[0] == "Norderney" or facilities[0] == "Bergen":
        for i in range(0,3):
            try:
                response = requests.get(url[i])
            except:
                print("error for Norderney or Bergen")
            data = json.loads(response.text)
            lastPoint = []
            for x, points in enumerate(data["prediction"][1]["trajectory"]):
                lastPoint = [points["datetime"],points["latitude"],points["longitude"],points["altitude"]]
            predicitons.append([facilities[0],data["request"]["launch_datetime"],lastPoint[0],lastPoint[1],lastPoint[2],lastPoint[3]])

for fligth in predicitons:
    radius = math.sqrt(math.pow((fligth[3]-homekoord[0]),2)+math.pow((fligth[4]-homekoord[1]),2))
    if radius <= 0.6:
        landings.append(fligth)
    else:
        notlandins.append(fligth)

print("""\
Subject: Eine Wettersonde ist gestartet! --automatisierte Benachrichtigung--

Folgende Sonden sind gestartet und werden in der Naehe von Habmburg landen:
""")

print("Report from:")
print(str(intoday))
print("Fligths that land in the given Region:")
if landings:
    linex = "============="
    for ziel in landings:
        line = "-----Launch facility: "+str(ziel[0])+"-----"
        linea = "Start Time: "+str(int(ziel[1][11:13])+1)+ "UHR"
        lineb = "ETA: "+str(int(ziel[2][11:13])+1)+":"+str(ziel[2][14:19])
        linec = "Landing Site: "+str(round(ziel[3],6))+"N "+str(round(ziel[4],6))+"O "
        lined = "last heigth: "+str(round(ziel[5],6))
        print('\n')
        print(line+ '\n')
        print(linea+ '\n')
        print(lineb+ '\n')
        print(linec+ '\n')
        print(lined+ '\n')
        print(linex+ '\n')
    print('\n')
    print('\n')
    print(linex+ '\n')
else:
    print("There are no landings in the given region")
print("")

print("Fligths that dont land in the given Region:")
if notlandins:
    linex = "============="
    for ziel in notlandins:
        line = "-----Launch facility: "+str(ziel[0])+"-----"
        linea = "Start Time: "+str(int(ziel[1][11:13])+1)+ "UHR"
        lineb = "ETA: "+str(int(ziel[2][11:13])+1)+":"+str(ziel[2][14:19])
        linec = "Landing Site: "+str(round(ziel[3],6))+"N "+str(round(ziel[4],6))+"O "
        lined = "last heigth: "+str(round(ziel[5],6))
        print('\n')
        print(line+ '\n')
        print(linea+ '\n')
        print(lineb+ '\n')
        print(linec+ '\n')
        print(lined+ '\n')
        print(linex+ '\n')
    print('\n')
    print('\n')
    print(linex+ '\n')
else:
    print("There are no landings outside the given region")
